### ISSUE TITLE
[HOLD for payment 2024-07-26][$500] When amending the amount, if the last digit is a 0, we don't show it.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#34894]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.29-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @twisterdotcom 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1705662590072039

## Action Performed:
1. create an expense report
2. Open the report and edit the amount so that last decimal digit is 0 (e.g. 23.20)
3. Again open and see the amount

## Expected Result:
Should see 0 at the end

## Actual Result:
"0" is missing

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/43996225/9bb584e7-49a7-4968-ab33-db9e71b281f8

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0178a9407eab9689fe</li>
        <li>Upwork Job ID: 1749451650573455360</li>
        <li>Last Price Increase: 2024-01-29</li>
<li>Automatic offers: </li>
<ul>
<li>abzokhattab | Contributor | 28129136</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @eVoloshchak</details>

**Comment by paultsimura** [comment]
This looks like a pretty normal behavior for decimals, isn't that expected?🤔

**Comment by cjoshi-zeals** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
If the trailing digit is 0 then it is getting truncated. Show 0 digits upto 2 decimals for amounts having significant number after the decimal
### What is the root cause of that problem?
in this function when dividing over 100 the last digit is removed if its 0
https://github.com/Expensify/App/blob/82b76c1a36cf1a01b00c14510a5812a591e7f2fd/src/libs/CurrencyUtils.ts#L91-L93

in the above function trailing 0 is getting truncated which is the root cause of the problem
``

### What changes do you think we should make in order to solve the problem?
We can make changes in the code such it will keep the trailing zero in case there is any significant digit after decimal and not in case it is a whole number which will be the edge case scenario.

So based on the bug if it is 23.10 it should display the same in amount input without truncating. 

and also in my solution for eg. 24.00 will be shown as 24 since trailing zeros after decimals don't have any significant digits.

keep the definition of the function intact without changing its return type.

```javascript
function convertToFrontendAmount(amountAsInt: number): number {
    // Convert the input integer to a floating-point number by dividing it by 100.0
    const amount = Math.trunc(amountAsInt) / 100.0;
    // Check if the decimal part is zero, then remove it
    if (amount % 1 === 0) {
        return amount;
    }
    // Return the resulting floating-point number
    return Number(amount.toFixed(2));
}
``` 

### What alternative solutions did you explore? (Optional)

**Comment by Aditya-svg-code** [comment]
### Proposal
**Please re-state the problem that we are trying to solve in this issue.**

if the last digit is a 0, it's unable to display the 0 after the decimal.

**What is the root cause of that problem?**

In the below function, if the amountAsInt has a number with unit digit 0, when dividing over 100.0 it ignores the 0 and displays only the remaining numbers.

For eg. if amountAsInt = 2370, then,  2370/100.0 = 23.7
```
function convertToFrontendAmount(amountAsInt: number): number {
    return Math.trunc(amountAsInt) / 100.0;
}
```

**What changes do you think we should make to solve the problem?**

When you perform a division operation, the result is often displayed with the minimum required precision. To format the result with a specific number of decimal places, you can use the toFixed() method.
toFixed() method returns a string, so if you need to perform further calculations, you may need to convert it back to a number using parseFloat() this function will return integer value. Here's how you can fix the issue.

```
function convertToFrontendAmount(amountAsInt: number): number {
    const result = Math.trunc(amountAsInt) / 100.0;
    return parseFloat(result.toFixed(2));
}
```

Additionally, if amountAsInt = 2300 then after division by 100.0 it became 23.0 only but in the above solution it will return 23.00 as result because in bills we have to show the numbers after the decimal even if it is 0.

**Comment by Aditya-svg-code** [comment]
### Proposal
**Please re-state the problem that we are trying to solve in this issue.**

when dividing the number by 100.0 it's unable to display the 0 after the decimal.

**What is the root cause of that problem?**

in this function when dividing over 100.0 the unit digit is removed if it is 0.
for eg. if amountAsInt =  2370 then 2370/100.0 = 23.7 but we want 23.70 



[App/src/libs/CurrencyUtils.ts](https://github.com/Expensify/App/blob/82b76c1a36cf1a01b00c14510a5812a591e7f2fd/src/libs/CurrencyUtils.ts#L91-L93)

Lines 91 to 93 in [82b76c1](https://github.com/Expensify/App/commit/82b76c1a36cf1a01b00c14510a5812a591e7f2fd)

```
 function convertToFrontendAmount(amountAsInt: number): number { 
     return Math.trunc(amountAsInt) / 100.0; 
 } 
```

**What changes do you think we should make to solve the problem?**

When you perform a division operation, the result is often displayed with the minimum required precision. To format the result with a specific number of decimal places, you can use the **toFixed()** method. 

toFixed() method returns a string, so if you need to perform further calculations, you may need to convert it back to a number using **parseFloat()**. Here's how you can fix the issue.

```
function convertToFrontendAmount(amountAsInt: number): number {
    const result = Math.trunc(amountAsInt) / 100.0;
    return parseFloat(result.toFixed(2));
}
```

Additionally, if amountAsInt = 2300 then the previous function was providing the result as 23.0 only, but in my solution it will return 23.00 as the result, usually in bills we should show the numbers after the decimal point it helps the customer understand the visibility of the payment.

**Comment by Aditya-svg-code** [comment]
Contributor details :
Expensify account email: [soni96996@gmail.com](mailto:soni96996@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~015162f5f4b61238ea

**Comment by twisterdotcom** [comment]
So, this was reported by me (which is why I added the label instead of taking it on), but I disagree that we shouldn't show the last `0`. I agree `12.30` is the same as `12.3`, but I found it odd that we didn't show the last character. If you asked your mate or your company to send you $12.30, you wouldn't ask them to send you $12.3.

**Comment by eVoloshchak** [comment]
I agree with @twisterdotcom, this doesn't look consistent
https://github.com/Expensify/App/assets/9059945/00e62fe4-c2b1-492e-9c34-d2495b486bb2
Banking apps also do this, if you put 12.3 it will still be displayed as 12.30 on the next step

@abzokhattab, I think this is the best [proposal](https://github.com/Expensify/App/issues/34894#issuecomment-1904253989) so far, but it has one problem - it changes the return type of `convertToFrontendAmount` from `number` to `string`. We need it to be number to calculate [isTaxAmountInvalid](https://github.com/Expensify/App/blob/90d54c0759f49691a02658285ec6d670d50cceeb/src/pages/iou/steps/MoneyRequestAmountForm.js#L71C7-L71C25)

@cjoshi-zeals, your proposal doesn't work for me, could you double-check it please?
> and also in my solution for eg. 24.00 will be shown as 24 since trailing zeros after decimals don't have any significant digits

We should display `24.00` in this case. 24 is displayed as 24.00 on the previous screen, those two should be identical

https://github.com/Expensify/App/assets/9059945/faedb657-a3cf-4fa0-8167-985d9d4b5037

@Aditya-svg-code, your [proposal](https://github.com/Expensify/App/issues/34894#issuecomment-1904635519) also doesn't work for me, could you double-check?

**Comment by cjoshi-zeals** [comment]
> We should display 24.00 in this case. 24 is displayed as 24.00 on the previous screen, those two should be identical

@eVoloshchak 

In that case, we just need to remove the condition for checking if there is no significant digit after the decimal, here is my updated solution

in the below solution will keep the return type of the function the same and will also deliver the desired output.

```javascript
function convertToFrontendAmount(amountAsInt: number): number {
    // Convert the input integer to a floating-point number by dividing it by 100.0
    const amount = Math.trunc(amountAsInt) / 100.0;
    
    // Return the resulting floating-point number
    return Number(amount.toFixed(2));
}
```

**Comment by Aditya-svg-code** [comment]
> I agree with @twisterdotcom, this doesn't look consistent https://github.com/Expensify/App/assets/9059945/00e62fe4-c2b1-492e-9c34-d2495b486bb2 Banking apps also do this, if you put 12.3 it will still be displayed as 12.30 on the next step
> 
> @abzokhattab, I think this is the best [proposal](https://github.com/Expensify/App/issues/34894#issuecomment-1904253989) so far, but it has one problem - it changes the return type of `convertToFrontendAmount` from `number` to `string`. We need it to be number to calculate [isTaxAmountInvalid](https://github.com/Expensify/App/blob/90d54c0759f49691a02658285ec6d670d50cceeb/src/pages/iou/steps/MoneyRequestAmountForm.js#L71C7-L71C25)
> 
> @cjoshi-zeals, your proposal doesn't work for me, could you double-check it please?
> 
> > and also in my solution for eg. 24.00 will be shown as 24 since trailing zeros after decimals don't have any significant digits
> 
> We should display `24.00` in this case. 24 is displayed as 24.00 on the previous screen, those two should be identical
> 
>  Screen.Recording.2024-01-24.at.23.57.22.mov 
> @Aditya-svg-code, your [proposal](https://github.com/Expensify/App/issues/34894#issuecomment-1904635519) also doesn't work for me, could you double-check?

@eVoloshchak I have double checked my solution and the issue is that converting it to the number and the leading zero after decimal is getting removed but if we keep it string then it is working correctly, and for the further calculation we will create a function `convertToFrontendAmountAsNumber()` that return a number for the calculation.

This is the function which will return string for the bill purpose.

```
function convertToFrontendAmount(amountAsInt: number): string { 
    const result = Math.trunc(amountAsInt) / 100.0;
    return (result.toFixed(2));              //return the value of result as string
}
```

This is the function which will return number for the furthur calculation.

```
 function convertToFrontendAmountAsNumber(amountAsInt: number): number { 
     return Math.trunc(amountAsInt) / 100.0; 
 } 
```

in `isTaxAmountInvalid` we need to use `convertToFrontendAmountAsNumber()` to get a number for calculation.

**Comment by sakluger** [comment]
I agree with @twisterdotcom - dollar amounts should always be reflected with two decimals.

**Comment by eVoloshchak** [comment]
> Instead of updating the currencyUtil, i propose that we update initializeAmount to fix the amount to 2 decimal places by using toFixed(2) as shown below:

@RohanSasne, this adds a small regression, amount is altered (formatted) when you're navigated to the next screen:

1. Plus button -> Request money
2. Enter an amount (for instance, 24)
3. Click 'Next'
4. Click 'Back'
5. Notice the amount is displayed as '24.00'

https://github.com/Expensify/App/assets/9059945/f1a66571-f02f-4f90-8b02-b193d7c49503

**Comment by dangrous** [comment]
Yeah I think @abzokhattab's is probably the clearest moving forward. @RohanSasne you're not wrong [here](https://github.com/Expensify/App/issues/34894#issuecomment-1914468621), it's not quite consistent, but I think if the user has JUST entered a value, we'd want to maintain it as is, making it easier for the user to fix if needed. (@sakluger or @twisterdotcom let me know if you disagree). This is also how Square Cash and Venmo do it, as a reference.

Two notes:
* Let's be sure to add a test for this in `CurrencyUtilsTest`
* We might want the new function to return an empty string if there's no value, so we can replace the full `amount ? CurrencyUtils.convertToFrontendAmount(amount).toString() : ''` line that's used, though I defer to your judgement on that.

**Comment by sakluger** [comment]
Hi @abzokhattab, how is the PR coming?

**Comment by dangrous** [comment]
hey @eVoloshchak @abzokhattab how are we looking? Seems like there's one bug we still need to fix, and then some merge conflicts. Can we get this ready for final review / merging by Monday or Tuesday? Thank you!

**Comment by dangrous** [comment]
Bumping this when you can, can we get this updated and rereviewed on Monday please? Thanks!

**Comment by dangrous** [comment]
How are we looking on this @abzokhattab ? Can we get it reviewed today, do you think?

**Comment by iwiznia** [comment]
Had to revert the PR due to https://github.com/Expensify/App/issues/39599

**Comment by eVoloshchak** [comment]
@abzokhattab, this bug was caused by a conflict when merging main, the commit at fault is https://github.com/Expensify/App/commit/bd6868207f1e8e03899515721ed6899eb50cf172

**Comment by dangrous** [comment]
Hrm okay let's see if we can figure out a new way forward that doesn't cause [that blocker](https://github.com/Expensify/App/issues/39599) (or https://github.com/Expensify/App/issues/39537)

**Comment by abzokhattab** [comment]
I will be provided a PR today. thanks

**Comment by dangrous** [comment]
How are we looking on the updated fix for this? Thanks!

**Comment by sakluger** [comment]
I removed the payment hold since the original PR caused a regression and we now have a new PR (https://github.com/Expensify/App/pull/40062) in review.

On that new PR, there is no BE reviewer. @eVoloshchak @dangrous is that correct?

**Comment by sakluger** [comment]
I'm also going to throw this in wave-collect. It could possibly be vsb, but I think collect is more appropriate.

**Comment by sakluger** [comment]
The [PR](https://github.com/Expensify/App/pull/40062) is still in review.

**Comment by sakluger** [comment]
@eVoloshchak it looks like we're waiting for your review on the PR. Could you please take a look?

**Comment by sakluger** [comment]
Still in review.

**Comment by sakluger** [comment]
@JmillsExpensify assigning you to help watch this issue while I'm OOO through May 31.

Currently, the [PR](https://github.com/Expensify/App/pull/40062) is in review. Can you please handle payment if needed? I'll take back over if it's still open when I return.

**Comment by JmillsExpensify** [comment]
Sounds good, thanks for the heads up.

**Comment by eVoloshchak** [comment]
Not overdue, final work is being done in https://github.com/Expensify/App/pull/40062

**Comment by sakluger** [comment]
Thanks for watching over this one @JmillsExpensify! I'm back and will take things back over.

**Comment by sakluger** [comment]
Looks like the PR is still in review.

**Comment by dangrous** [comment]
One of these ^ I don't think is a bug, just new behavior, but https://github.com/Expensify/App/issues/43215 feels real and fixable

**Comment by sakluger** [comment]
@luacmartins @dangrous has [this PR](https://github.com/Expensify/App/pull/40062) been deployed yet? I don't see any deploy comment, and am wondering how it cause those regressions if it hasn't been deployed.

**Comment by luacmartins** [comment]
Yes, the PR was deployed and included in this checklist. https://github.com/Expensify/App/issues/43168

**Comment by sakluger** [comment]
Sorry, I'm a bit confused. I think that we still have https://github.com/Expensify/App/issues/43215 as a regression stemming from this issue, so we should hold payment until that is fixed. @dangrous @eVoloshchak could you please confirm if that's correct?

**Comment by dangrous** [comment]
Yep that's correct - the bot does not do well with regressions.

We're waiting on review/merge of https://github.com/Expensify/App/pull/43816 which should solve #43215

**Comment by sakluger** [comment]
Thanks @dangrous for clarifying! I've updated the title to be a bit more clear, and I'm going to move this back to weekly while we wait for https://github.com/Expensify/App/pull/43816 to be merged & deployed.

**Comment by sakluger** [comment]
The PR is still not merged.

**Comment by dangrous** [comment]
PR in review! Hopefully should get merged today

**Comment by dangrous** [comment]
we're merged!

**Comment by dangrous** [comment]
No automation for some reason, but this has been merged and deployed. Should be ready for payment on 7/26

**Comment by sakluger** [comment]
Thanks @dangrous 👍 - I updated the title accordingly and will come back tomorrow to make payments.

**Comment by sakluger** [comment]
Summarizing payment on this issue:

Contributor: @abzokhattab $500, paid via Upwork
Contributor+: @eVoloshchak $500, please request on Newdot

**Comment by sakluger** [comment]
@eVoloshchak can you please complete the BZ checklist?

**Comment by sakluger** [comment]
Bump @eVoloshchak

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: this was the original implementation
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A

**Regression Test Proposal**

1. Click on FAB
2. Select Request Money -> Manual tab
3. Enter an amount that has 0 as the last decimal digit (e.g. 23.20)
4. Click on Next > go back using the header back button
5. Ensure the amount shown to the user is the same as in step 4 (the trailing zero isn't cut off)

**Do we agree 👍 or 👎**

**Comment by sakluger** [comment]
Thanks!
